WordPress/Sniffs/NamingConventions/ValidPostTypeSlugSniff.php

```diff
@@ -0,0 +1,208 @@
+		'wp_block'            => true,
+		'action'              => true,
+		'author'              => true,
+		'order'               => true,
+		'theme'               => true,
+	);
+
+	/**
+	 * All valid tokens for in the first parameter of register_post_type().
+	 *
+	 * Set in `register()`.
+	 *
+	 * @var string
+	 */
+	public $valid_tokens = array();
+
+	/**
+	 * Returns an array of tokens this test wants to listen for.
+	 *
+	 * @return array
+	 */
+	public function register() {
+		$this->valid_tokens = Tokens::$textStringTokens + Tokens::$heredocTokens;
+		return parent::register();
+	}
+
+	/**
+	 * Process the parameter of a matched function.
+	 *
+	 * Errors on invalid post type names when reserved keywords are used,
+	 * the post type is too long, or contains invalid characters.
+	 *
+	 * @since X.X.X
+	 *
+	 * @param int    $stackPtr        The position of the current token in the stack.
+	 * @param array  $group_name      The name of the group which was matched.
+	 * @param string $matched_content The token content (function name) which was matched.
+	 * @param array  $parameters      Array with information about the parameters.
+	 *
+	 * @return void
+	 */
+	public function process_parameters( $stackPtr, $group_name, $matched_content, $parameters ) {
+		// Check for non string based slug parameter (we cannot determine if this is valid).
+		$contains_text_string_token = $this->phpcsFile->findNext( $this->valid_tokens, $parameters[1]['start'], $parameters[1]['end'] + 1 );
+		if ( ! $contains_text_string_token || ( $contains_text_string_token && $this->phpcsFile->findNext( Tokens::$functionNameTokens, $parameters[1]['start'], $parameters[1]['end'] + 1 ) !== false ) ) {
+			$this->phpcsFile->addWarning(
+				'Dynamic post type slug found. It is not possible to determine the validity of this slug. Found: %s.',
+				$stackPtr,
+				'Dynamic',
+				array(
+					$parameters[1]['raw'],
+				),
+				3
+			);
+			return;
+		}
+
+		// Handling string interpolation.
+		$stringPos = $this->phpcsFile->findNext( Tokens::$textStringTokens, $parameters[1]['start'], $parameters[1]['end'] + 1, false, null, true );
+		if ( false === $stringPos ) {
+			return;
+		}
+
+		$post_type = $this->strip_quotes( $this->tokens[ $stringPos ]['content'] );
+		$data      = array(
+			$this->tokens[ $stringPos ]['content'],
+		);
+
+		// Warn for dynamic parts in the slug parameter.
+		if ( 'T_DOUBLE_QUOTED_STRING' === $this->tokens[ $stringPos ]['type'] || ( 'T_HEREDOC' === $this->tokens[ $stringPos ]['type'] && strpos( $this->tokens[ $stringPos ]['content'], '$' ) !== false ) ) {
+			$this->phpcsFile->addWarning(
+				'register_post_type() slug may or may not get too long with dynamic contents and could contain invalid characters. Found: %s.',
+				$stringPos,
+				'PartiallyDynamic',
+				$data
+			);
+			$post_type = $this->strip_interpolated_variables( $post_type );
+		}
+
+		// Error for using reserved slug names.
+		if ( strlen( $post_type ) === 0 ) {
```

Comment:
As all the other checks are irrelevant when the post type is empty, I'd like to suggest moving this up to above the "dynamic contents" check and adding a `return` to it.
For this specific case, we know that none of the other errors can happen if this error gets triggered, so a `return` here after the error is thrown, is perfectly fine.

Reply:
Changing this back to "Unresolved".

Sorry, my previous remark may not have been clear enough, I meant _moving this check up above the `PartiallyDynamic`_ error, not to above the `NotStringLiteral` error.

This check has now been moved too far up, will never be triggered and isn't tested either.

If the function call doesn't have any parameter, the `process_parameters() method won't be called, the `process_no_parameters()` method will be called instead.
See: https://github.com/WordPress/WordPress-Coding-Standards/blob/058be1bc5b2c5f20bcb0fb44603e43abe6315be4/WordPress/AbstractFunctionParameterSniff.php#L69-L78

And if it's an empty string, the check still will not be triggered as it is now above the point in the sniff where `$this->strip_quotes()` is called, so the string length wouldn't be `0`, but `2`.

You can verify this by running the sniff against the unit tests with `--report=full,source`, the `Empty` code is not triggered by the tests.

Reply:
Right, didn't know about the `process_no_parameters()`, and was thinking about catching the empty case as early as possible.

It now actually correctly checks for an empty first param after the latest commit.